### PR TITLE
bpo-38821: Fix crash in argparse when using gettext

### PR DESCRIPTION
**Pull request by federicobond**
https://bugs.python.org/issue38821

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@federicobond

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by brandtbucher** [comment]
Thanks for your time, and welcome to CPython @federicobond!

I assume that you've seen the bot's message about the CLA?

**Comment by federicobond** [comment]
Hi there @brandtbucher, yes. Waiting for the CLA to get accepted! Should I include a changelog in the PR?

**Comment by brandtbucher** [comment]
Yes please! You can do so easily [here](https://blurb-it.herokuapp.com).

**Comment by federicobond** [comment]
Done! I was not feeling particularly inspired writing the news entry, so feel free to suggest changes.

**Comment by rhettinger** [comment]
Please add a test.

**Comment by federicobond** [comment]
I could not find any existing test exercising locale support in argparse. What would this test assert?

**Comment by taleinat** [comment]
> I could not find any existing test exercising locale support in argparse. What would this test assert?

I suggest test various possible values for nargs, including all of the special ones, when using gettext.

**Comment by federicobond** [comment]
I think adding those tests would be a great idea but none of them would fail prior to this change, so I am not sure we should bundle them together.

A proper regression test for this change needs to load a custom catalog, bind it to the global domain, etc.

File: Misc/NEWS.d/next/Library/2019-11-16-23-26-25.bpo-38821.-albNN.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix crash in argparse when using gettext.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Library/2019-11-16-23-26-25.bpo-38821.-albNN.rst]:
```suggestion
Fix unhandled exceptions in :mod:`argparse` when internationalizing some error messages.
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Just a bit of clarification on the NEWS entry:

**Comment by taleinat** [comment]
> A proper regression test for this change needs to load a custom catalog, bind it to the global domain, etc.

We could also monkey-patch `argparse._` and `argparse.ngettext` using `unittest.mock.patch`, which should be simpler to set up.

That being said, this is an obvious fix and we have no existing tests for this, so I realize demanding that you add such tests is rather unfair. Let me know whether you'd like to try, and I'll take care of things even if you don't.

**Comment by federicobond** [comment]
@taleinat sure, I can do that. Do you think those tests should accompany this PR or should I open a new one? I would rather submit another PR so we can discuss them separately and not stall this changeset, but whatever works best for the maintainers is good for me.

**Comment by taleinat** [comment]
@federicobond, yes, let's go with a separate issue and PR for adding such tests.

File: Misc/NEWS.d/next/Library/2019-11-16-23-26-25.bpo-38821.-albNN.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix unhandled exceptions in :mod:`argparse` when internationalizing some error messages.
```
Comment (taleinat) [inline Misc/NEWS.d/next/Library/2019-11-16-23-26-25.bpo-38821.-albNN.rst]:
```suggestion
Fix unhandled exceptions in :mod:`argparse` when internationalizing error messages for arguments with ``nargs`` set to special (non-integer) values.  Patch by Federico Bond.
```

**Review by taleinat — APPROVED** [review]

**Comment by brandtbucher** [comment]
Congrats on your first CPython contribution @federicobond! :champagne:

Looking forward to seeing more from you in the future.

**Comment by federicobond** [comment]
Thank you @brandtbucher & co! I can see that you have all put a lot of effort into making this a pleasurable experience, definitely makes me want to contribute again 🎉

**Pull request merged**